### ISSUE TITLE
getPreviousNode() behaves differently

### ISSUE THREAD

**Issue by odungern** [mbraak/jqTree#786]
v1.8.0: This is just a note, not an issue for us (as we have already changed our algorithms):

Consider a tree:
• Node-1
• Node-2
&nbsp;&nbsp;• Node-2.1
&nbsp;&nbsp;• Node-2.2
&nbsp;&nbsp;&nbsp;&nbsp;• Node-2.2.1
&nbsp;&nbsp;&nbsp;&nbsp;• Node-2.2.2
• Node-3 <-- selected
• Node-4

With Node-3 selected:
- If both Node-2 and Node-2.2 are closed, getPreviousNode now returns Node-2.2, whereas in earlier releases it was Node-2.2.2.
- If Node-2 is closed and Node-2.2 is open, getPreviousNode returns Node-2.2.2 as before.
- If Node-2 is open and Node-2.2 is closed, getPreviousNode now returns Node-2.2, whereas in earlier releases it was Node-2.2.2.
- If both Node-2 and Node-2.2 are open, getPreviousNode returns Node-2.2.2 as before.

In other words, before the node returned by getPreviousNode was always Node-2.2.2, whereas now it depends on whether Node-2.2 is open or closed.

As I have said, no problem for us ...

**Comment by mbraak** [comment]
I can't reproduce this error. When I try to reproduce it, I get the following:

* The method `getPreviousNode` always returns the node 2.2.2., regardless if nodes are open or closed.
* The method `getPreviousVisibleNode` does depend on nodes being open or closed.
  * If node 2.2 is open, it returns node 2.2.2 (because that node is visible)
  * If node 2.2 is closed, it returns node 2.2 (because node 2.2.2 is not visible)

I'm happy that it works for you, but I don't understand why we're getting different results. 

Note that the behaviour of `getPreviousNode` and `getNextNode` was changed in version 1.7.0. In that version also the methods `getPreviousVisibleNode` and `getNextVisibleNode` were added. Also see the [changelog](https://mbraak.github.io/jqTree/#changelog).

**Comment by odungern** [comment]
Thanks for having a look into it, right away, Marco. But unfortunately, I cannot confirm your observations.

We have tried again (and we hope that our adding attributes to the tree doesn't change it behavior - we did not change any existing attributes, though).

Again in the example above, lets assume that all nodes are closed and Node-3 is selected:
- getPreviousSibling() returns Node-2 (as expected)
- getPreviousVisibleNode() returns Node-2 (as expected)
- getPreviousNode() delivers Node-2.2 (expected: Node-2.2.2)
- with other opened/closed configurations we have verified the behavior as described in the original post.
- getNextSibling() returns Node-3 (as expected)
- getNextVisibleNode() returns Node-3 (as expected)
- getNextNode() delivers Node-3.1 (as expected)
- if Node-3 is open, then getNextVisibleNode() returns Node-3.1 (as expected), the return value of the other two functions is unchanged.

No problem for us, as we have already changed our algorithms based on the new behavior. Up to you, if you consider it right or wrong. From our perspective, though, it would be more logical, if getPreviousNode() returns the last child having no children as deep as it gets - and independently of whether any of the nodes is open or closed.

**Comment by mbraak** [comment]
Thanks for looking into. And no, adding attributes doesn't change the behavior. The only important attributes are the id and the name.

I will look into this later. Closing the issue for now, because it's no direct problem for you.

**Comment by odungern** [comment]
Just for the records:

Assuming that all nodes are open and Node-2.2.1 selected:
- getPreviousSibling() returns null (as expected)
- getPreviousVisibleNode() returns Node-2.2. (as expected)
- getPreviousNode() returns Node-2.2. (as expected)

Assuming that all nodes are open and Node-2.2.2 selected:
- getNextSibling() returns null (as expected)
- getNextVisibleNode() returns Node-3 (as expected)
- getNextNode() returns Node-3 (as expected)

Also for the records: The jqtree is a very useful, well designed and well implemented piece of code. Thanks a lot for your effort, Marco. Our project [SpecIF Model Integrator and Editor](https://github.com/GfSE/SpecIF-Viewer) is also benevolent, so unfortunately no revenues to share ;-)
